<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simulation of genetic values based on an unstructured model for genotype-by-environment (GxE) interaction • FieldSimR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulation of genetic values based on an unstructured model for genotype-by-environment (GxE) interaction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FieldSimR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/compound_symmetry_GxE_demo.html">Simulation of genetic values based on a compound symmetry model for genotype-by-environment (GxE) interaction</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_variation_demo.html">Simulation of plot errors and phenotypes in a plant breeding field trial</a></li>
    <li><a class="dropdown-item" href="../articles/unstructured_GxE_demo.html">Simulation of genetic values based on an unstructured model for genotype-by-environment (GxE) interaction</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/crWerner/fieldsimr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulation of genetic values based on an unstructured model for genotype-by-environment (GxE) interaction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/crWerner/fieldsimr/blob/main/vignettes/unstructured_GxE_demo.Rmd" class="external-link"><code>vignettes/unstructured_GxE_demo.Rmd</code></a></small>
      <div class="d-none name"><code>unstructured_GxE_demo.Rmd</code></div>
    </div>

    
    
<p><br></p>
<div class="section level3">
<h3 id="background">
<strong>Background</strong><a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p>This document demonstrates how to simulate genetic values for
multiple traits in multiple environments based on an
<strong>unstructured model for genotype-by-environment (GxE)
interaction</strong>. While the simulation of genetic values is not
directly implemented in <a href="https://cran.r-project.org/package=FieldSimR" class="external-link">‘FieldSimR’</a>, the
package provides wrapper functions to facilitate the simulation of
genetic values for multi-environment field trial settings employing the
R package <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a>. The
two wrapper functions to simulate genetic values based on an
unstructured model for GxE interaction are:</p>
<ul>
<li><code><a href="../reference/unstr_asr_input.html">unstr_asr_input()</a></code></li>
<li><code><a href="../reference/unstr_asr_output.html">unstr_asr_output()</a></code></li>
</ul>
<p><br></p>
<p><strong>Note:</strong> <a href="https://cran.r-project.org/package=FieldSimR" class="external-link">‘FieldSimR’</a> also
provides wrapper functions to enable simulation of genetic values using
a compound symmetry model for GxE interaction. This is demonstrated in
the vignette on the <a href="https://crwerner.github.io/fieldsimr/articles/compound_symmetry_GxE_demo.html">Simulation
of genetic values based on a compound symmetry model for
genotype-by-environment (GxE) interaction</a>.</p>
<p>The <strong>core function of <a href="https://cran.r-project.org/package=FieldSimR" class="external-link">‘FieldSimR’</a></strong>
generates plot errors comprising 1) a spatially correlated error term,
2) a random error term, and 3) an extraneous error term. Spatially
correlated errors are simulated using either bivariate interpolation, or
a two-dimensional autoregressive process of order one (AR1:AR1). Through
the combination of the plot errors with (simulated) genetic values, <a href="https://cran.r-project.org/package=FieldSimR" class="external-link">‘FieldSimR’</a>
enables simulation of multi-environment plant breeding trials at the
plot. This is demonstrated in the vignette on the <a href="https://crwerner.github.io/fieldsimr/articles/spatial_variation_demo.html">Simulation
of plot errors and phenotypes in a plant breeding field trial</a>.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="simulation-of-genetic-values">
<strong>Simulation of genetic values</strong><a class="anchor" aria-label="anchor" href="#simulation-of-genetic-values"></a>
</h3>
<p>We conceive a scenario in which 100 maize hybrids are measured for
grain yield (t/ha) and plant height (cm) in three environments. The
first and the third location include two replicates, and the third
location includes three replicates.</p>
<p><strong>The simulation process comprises three steps:</strong></p>
<ol style="list-style-type: decimal">
<li>Definition of the genetic architecture and the simulation parameters
for the two traits.</li>
<li>Simulation of a population of 100 hybrid genotypes.</li>
<li>Generation of a data frame containing the simulated genetic values
for grain yield and plant height in the three environments.</li>
</ol>
<p>To provide a comprehensive overview of the unstructured modelling
approach for GxE interaction, we assume additive and dominance gene
action for both grain yield and plant height. Details on how <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a>
simulates additive and non-additive biological effects can be found in
the <a href="https://cran.r-project.org/package=AlphaSimR/vignettes/traits.pdf" class="external-link">“Traits
in AlphaSimR”</a> vignette.</p>
<p>It should be noted, however, that a simple additive genetic model
will be sufficient to answer most experimental questions and may be
preferred to more complex models, especially if data to tune the
simulation model is not available and the parameters are unknown.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="genetic-architecture-and-simulation-parameters-of-the-two-traits">
<strong>1. Genetic architecture and simulation parameters of the two
traits</strong><a class="anchor" aria-label="anchor" href="#genetic-architecture-and-simulation-parameters-of-the-two-traits"></a>
</h3>
<p>First, we set the number of traits, the number of environments (e.g.,
locations), and the number of replicates tested within environments. We
also define the number of genotypes in our founder population to be
simulated, the number of chromosomes, and the number of segregating
sites (biallelic QTN) per chromosome. Then, we set the additive genetic
parameters, the dominance parameters, and the genetic correlation
structures required to simulate the two traits in three environments
based on an unstructured model for GxE interaction.</p>
<p>We create a founder population of 20 heterozygous genotypes. These
founder genotypes will then be split into two heterotic pools, and one
doubled haploid (DH) line will be produced from each founder. To
generate hybrids, the two pools will be crossed using a factorial
design.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ntraits</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># Number of traits.</span></span>
<span><span class="va">nenvs</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># Number of environments.</span></span>
<span><span class="va">nreps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span> <span class="co"># Number of replicates tested within environments 1, 2 and 3.</span></span>
<span></span>
<span></span>
<span><span class="va">nind</span> <span class="op">&lt;-</span> <span class="fl">20</span> <span class="co"># Number of founder genotypes in the population.</span></span>
<span><span class="va">nchr</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co"># Number of chromosomes.</span></span>
<span><span class="va">nseg_sites</span> <span class="op">&lt;-</span> <span class="fl">200</span> <span class="co"># Number of QTN per chromosome.</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="additive-genetic-parameters">
<strong>Additive genetic parameters</strong><a class="anchor" aria-label="anchor" href="#additive-genetic-parameters"></a>
</h4>
<p>We define mean additive genetic values for all <strong><em>trait x
environment</em></strong> combinations. The additive mean values are
provided in a single vector with environments nested within traits.
Grain yield is measured in tons per hectare (t/ha) and plant height is
measured in centimetres (cm).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4.9</span>, <span class="fl">5.4</span>, <span class="fl">5.1</span>, <span class="fl">235.2</span>, <span class="fl">228.5</span>, <span class="fl">239.1</span><span class="op">)</span> <span class="co"># c(Yld:E1, Yld:E2, Yld:E3, Prt:E1, Prt:E2, Prt:E3)</span></span></code></pre></div>
<p><br></p>
<p>The genetic variances for the six <strong><em>trait x
environment</em></strong> combinations can be set in two different
ways.</p>
<ol style="list-style-type: decimal">
<li>By providing a single vector with environments nested within traits
(similar to vector <code>mean</code>).</li>
<li>By providing a vector of genetic variances for the two traits
(<code>T_var</code>) and a second vector of genetic variances for the
three environments (<code>E_var</code>), assuming a separable
structure.</li>
</ol>
<p>We choose the first option and provide a single vector of
variances.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.085</span>, <span class="fl">0.12</span>, <span class="fl">0.06</span>, <span class="fl">15.1</span>, <span class="fl">8.5</span>, <span class="fl">11.7</span><span class="op">)</span> <span class="co"># c(Yld:E1, Yld:E2, Yld:E3, Pht:E1, Pht:E2, Pht:E3)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="dominance-genetic-parameters">
<strong>Dominance genetic parameters</strong><a class="anchor" aria-label="anchor" href="#dominance-genetic-parameters"></a>
</h4>
<p>We assume mean dominance degrees of 0.4 for grain yield and 0.1 for
plant height in all three environments, respectively. The dominance
degree variance is set to 0.2 for both traits in the three
environments.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meanDD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span> <span class="co"># c(Yld:E1, Yld:E2, Yld:E3, Pht:E1, Pht:E2, Pht:E3)</span></span>
<span><span class="va">varDD</span> <span class="op">&lt;-</span> <span class="fl">0.2</span> <span class="co"># Same value set for all environment-within-trait combinations</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="genetic-correlation-structures">
<strong>Genetic correlation structures</strong><a class="anchor" aria-label="anchor" href="#genetic-correlation-structures"></a>
</h4>
<p>Additive and non-additive genetic correlations between all
<strong><em>trait x environment</em></strong> combinations can be
defined in two ways:</p>
<ol style="list-style-type: decimal">
<li>By providing a single correlation structure
(<code>corA</code>).</li>
<li>By providing a matrix of genetic correlations between traits
(<code>TcorA</code>) and a matrix of genetic correlations between
environments (<code>EcorA</code>), assuming a separable correlation
structure.</li>
</ol>
<p>We choose the second option and assume a <strong>separable additive
genetic correlation structure</strong>. This simplifies the generation
of the additive correlation matrix, especially when the number of
environments and/or traits is high.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TcorA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span> <span class="co"># Matrix of additive genetic correlations between the two traits.</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fl">1.0</span>, <span class="fl">0.6</span>,</span>
<span>    <span class="fl">0.6</span>, <span class="fl">1.0</span></span>
<span>  <span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;      [,1] [,2]</span></span>
<span><span class="co">#&gt; [1,]  1.0  0.6</span></span>
<span><span class="co">#&gt; [2,]  0.6  1.0</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EcorA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="co"># Matrix of additive genetic correlations between the three environments.</span></span>
<span>    <span class="fl">1.0</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>,</span>
<span>    <span class="fl">0.4</span>, <span class="fl">1.0</span>, <span class="fl">0.5</span>,</span>
<span>    <span class="fl">0.6</span>, <span class="fl">0.5</span>, <span class="fl">1.0</span></span>
<span>  <span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;      [,1] [,2] [,3]</span></span>
<span><span class="co">#&gt; [1,]  1.0  0.4  0.6</span></span>
<span><span class="co">#&gt; [2,]  0.4  1.0  0.5</span></span>
<span><span class="co">#&gt; [3,]  0.6  0.5  1.0</span></span></code></pre>
<p><br></p>
<p>Alternatively, the function <code><a href="../reference/rand_cor_mat.html">rand_cor_mat()</a></code> provides a
quick means to generate random correlation matrices. The correlations
are sampled within a defined range of values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rand_cor_mat.html">rand_cor_mat</a></span><span class="op">(</span> <span class="co"># Additive genetic correlation structure.</span></span>
<span>  <span class="op">(</span><span class="va">ntraits</span> <span class="op">*</span> <span class="va">nenvs</span><span class="op">)</span>, <span class="co"># Could be used instead of TcorA and EcorA.</span></span>
<span>  min.cor <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  max.cor <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  pos.def <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">corA</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;      1    2    3    4    5    6</span></span>
<span><span class="co">#&gt; 1 1.00 0.17 0.74 0.59 0.25 0.10</span></span>
<span><span class="co">#&gt; 2 0.17 1.00 0.46 0.50 0.34 0.68</span></span>
<span><span class="co">#&gt; 3 0.74 0.46 1.00 0.68 0.74 0.27</span></span>
<span><span class="co">#&gt; 4 0.59 0.50 0.68 1.00 0.15 0.34</span></span>
<span><span class="co">#&gt; 5 0.25 0.34 0.74 0.15 1.00 0.40</span></span>
<span><span class="co">#&gt; 6 0.10 0.68 0.27 0.34 0.40 1.00</span></span></code></pre></div>
<p><br></p>
<p>Furthermore, we assume <strong>independence between the dominance
degrees</strong> of the six
<strong><em>trait-by-environment</em></strong> combinations. Therefore,
we generate a 6x6 diagonal matrix (although this is not strictly
necessary. A diagonal matrix is constructed by default if no correlation
matrix is provided).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corDD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6]</span></span>
<span><span class="co">#&gt; [1,]    1    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [2,]    0    1    0    0    0    0</span></span>
<span><span class="co">#&gt; [3,]    0    0    1    0    0    0</span></span>
<span><span class="co">#&gt; [4,]    0    0    0    1    0    0</span></span>
<span><span class="co">#&gt; [5,]    0    0    0    0    1    0</span></span>
<span><span class="co">#&gt; [6,]    0    0    0    0    0    1</span></span></code></pre>
<p><br></p>
</div>
<div class="section level4">
<h4 id="input-parameter-list">
<strong>Input parameter list</strong><a class="anchor" aria-label="anchor" href="#input-parameter-list"></a>
</h4>
<p>Once we have defined all simulation parameters, we use the function
<code><a href="../reference/unstr_asr_input.html">unstr_asr_input()</a></code> to prepare them into a list that is used
with <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a> to
simulate correlated genetic values based on an unstructured model for
GxE interaction.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_asr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unstr_asr_input.html">unstr_asr_input</a></span><span class="op">(</span></span>
<span>  ntraits <span class="op">=</span> <span class="va">ntraits</span>,</span>
<span>  nenvs <span class="op">=</span> <span class="va">nenvs</span>,</span>
<span>  mean <span class="op">=</span> <span class="va">mean</span>,</span>
<span>  var <span class="op">=</span> <span class="va">var</span>,</span>
<span>  TcorA <span class="op">=</span> <span class="va">TcorA</span>,</span>
<span>  EcorA <span class="op">=</span> <span class="va">EcorA</span>,</span>
<span>  meanDD <span class="op">=</span> <span class="va">meanDD</span>,</span>
<span>  varDD <span class="op">=</span> <span class="va">varDD</span>,</span>
<span>  corDD <span class="op">=</span> <span class="va">corDD</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> The object <code>input_asr</code> should not
be modified and must be used directly with <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a> as
demonstrated below.</p>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="simulation-of-a-population-of-genotypes">
<strong>2. Simulation of a population of genotypes</strong><a class="anchor" aria-label="anchor" href="#simulation-of-a-population-of-genotypes"></a>
</h3>
<p>Our list of simulation parameters <code>input_asr</code> is now used
with <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a> to
simulate correlated genetic values for 100 maize hybrid genotypes tested
for two traits in three environments based on an unstructured model for
GxE interaction.</p>
<p>First, we <strong>simulate a population of 20 heterozygous maize
founder genotypes</strong> using the function <code>runMacs</code> in <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">founders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/runMacs.html" class="external-link">runMacs</a></span><span class="op">(</span> <span class="co"># Simulation of founder genotypes using AlphaSimR's "MAIZE" presets</span></span>
<span>  nInd <span class="op">=</span> <span class="va">nind</span>, <span class="co"># to mimic the species' evolutionary history.</span></span>
<span>  nChr <span class="op">=</span> <span class="va">nchr</span>,</span>
<span>  segSites <span class="op">=</span> <span class="va">nseg_sites</span>,</span>
<span>  inbred <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"MAIZE"</span>,</span>
<span>  nThreads <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">SP</span> <span class="op">&lt;-</span> <span class="va"><a href="https://gaynorr.github.io/AlphaSimR/reference/SimParam.html" class="external-link">SimParam</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">founders</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>Then, we use the simulation parameters stored in
<code>input_asr</code> to simulate <strong>correlated genetic values for
grain yield and plant height in three testing environments</strong>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SP</span><span class="op">$</span><span class="fu">addTraitAD</span><span class="op">(</span> <span class="co"># Additive + dominance trait simulation.</span></span>
<span>  nQtlPerChr <span class="op">=</span> <span class="va">nseg_sites</span>,</span>
<span>  mean <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">mean</span>,</span>
<span>  var <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">var</span>,</span>
<span>  corA <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">corA</span>,</span>
<span>  meanDD <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">meanDD</span>,</span>
<span>  varDD <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">varDD</span>,</span>
<span>  corDD <span class="op">=</span> <span class="va">input_asr</span><span class="op">$</span><span class="va">corDD</span>,</span>
<span>  useVarA <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">founders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/newPop.html" class="external-link">newPop</a></span><span class="op">(</span><span class="va">founders</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>We now <strong>split the simulated founders into two heterotic pools
A and B</strong>. We create one DH line per founder, which gives us 10
DH lines per heterotic pool. Hybrids are then generated by crossing pool
A and pool B in a factorial manner (all pairwise combinations),
resulting in 100 hybrid genotypes</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pool_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/makeDH.html" class="external-link">makeDH</a></span><span class="op">(</span><span class="va">founders</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, nDH <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># Pool A: 1 DH line from founders 1 to 10, respectively.</span></span>
<span><span class="va">pool_B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/makeDH.html" class="external-link">makeDH</a></span><span class="op">(</span><span class="va">founders</span><span class="op">[</span><span class="fl">11</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, nDH <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># Pool B: 1 DH line from founders 11 to 20, respectively.</span></span>
<span></span>
<span><span class="va">dh_lines</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/mergePops.html" class="external-link">mergePops</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pool_A</span>, <span class="va">pool_B</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">factorial_plan</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">pool_A</span><span class="op">@</span><span class="va">id</span>, B <span class="op">=</span> <span class="va">pool_B</span><span class="op">@</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="co"># Factorial crossing plan.</span></span>
<span></span>
<span><span class="va">hybrid_pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/makeCross.html" class="external-link">makeCross</a></span><span class="op">(</span>pop <span class="op">=</span> <span class="va">dh_lines</span>, crossPlan <span class="op">=</span> <span class="va">factorial_plan</span>, nProgeny <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># Hybrid genotypes.</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="generation-of-a-data-frame-with-simulated-genetic-values">
<strong>3. Generation of a data frame with simulated genetic
values</strong><a class="anchor" aria-label="anchor" href="#generation-of-a-data-frame-with-simulated-genetic-values"></a>
</h3>
<p>In the last step, we use the function <code><a href="../reference/unstr_asr_output.html">unstr_asr_output()</a></code>
to extract the simulated genetic values from the <a href="https://CRAN.R-project.org/package=AlphaSimR" class="external-link">‘AlphaSimR’</a>
population object <code>hybrid_pop</code> and store them in a data
frame.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gv_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unstr_asr_output.html">unstr_asr_output</a></span><span class="op">(</span></span>
<span>  pop <span class="op">=</span> <span class="va">hybrid_pop</span>,</span>
<span>  ntraits <span class="op">=</span> <span class="va">ntraits</span>,</span>
<span>  nenvs <span class="op">=</span> <span class="va">nenvs</span>,</span>
<span>  nreps <span class="op">=</span> <span class="va">nreps</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   env id rep gv.Trait1 gv.Trait2</span></span>
<span><span class="co">#&gt; 1   1 41   1  5.135440  241.6967</span></span>
<span><span class="co">#&gt; 2   1 42   1  4.941166  242.0755</span></span>
<span><span class="co">#&gt; 3   1 43   1  4.963438  238.0766</span></span>
<span><span class="co">#&gt; 4   1 44   1  4.911714  237.0087</span></span>
<span><span class="co">#&gt; 5   1 45   1  5.108014  238.7757</span></span>
<span><span class="co">#&gt; 6   1 46   1  4.931817  233.8303</span></span></code></pre>
<p><br></p>
<p><strong>Histogram showing the genetic values of the 100 maize hybrids
for grain yield in the three environments</strong></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">gv_df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">gv.Trait1</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">env</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#e9ecef"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, position <span class="op">=</span> <span class="st">"identity"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"violetred3"</span>, <span class="st">"goldenrod3"</span>, <span class="st">"skyblue2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Genetic values for grain yield (t/ha)"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, fill <span class="op">=</span> <span class="st">"Environment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="unstructured_GxE_demo_files/figure-html/unnamed-chunk-19-1.png" width="864" style="display: block; margin: auto;"></p>
<p><br></p>
<p>The simulated genetic values for grain yield and plant height
measured at three environments can now be combined with plot errors to
generate plant breeding trial phenotype plot data. We provide an example
on how to do that in the vignette <a href="https://crwerner.github.io/fieldsimr/articles/spatial_variation_demo.html">Simulation
of plot errors and phenotypes in a plant breeding field trial</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Christian Werner, Daniel Tolhurst.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
